README.md
# platsbanken-vacancy
Submit vacancies to the Arbetsförmedlingen Platsbanken (Swedish employment agency
job centre).

## Description
This module provides an interface for generating an appropriate XML document for
submission to the Arbetsförmedlingen Platsbanken API.

It does not perform the submission or handle responses. See `./example/example.js`.

## Install

`npm install --save @othermachines/platsbanken-vacancy`

## Scripts

npm test: `mocha`

npm run build: `babel src --out-dir build --source-maps`

npm run watch: `babel src --out-dir build --source-maps --watch`

npm prepare: `npm run build`

## Dependencies
This module was developed against node v8.1.2.

- joi
- source-map-support
- xml

## Usage

```
const vacancy = require('@othermachines/platsbanken-vacancy');

const request = vacancy();

try {
  request
    .sender({
      id: '12345678',
      email: 'foo@bar.com',
    })
    .transaction({
      id: 'TRANSACTION GUID',
    })
    .jobPositionPosting({
      id: '46-123456-1234-123-123',
      status: 'active',
    })
    .hiringOrg({
      name: 'Company Name',
      id: '46-123456-1234'
      url: 'http://example.org',
      description: 'Organizational unit description',
    })
    .hiringOrgContact({
      countryCode: 'SE',
      postalCode: '11356',
      municipality: '0180',
      addressLine: 'Birger Jarlsgatan 58, 11356, Stockholm',
      streetName: 'Birger Jarlsgatan 58',
    })
    .postDetail({
      startDate: '2018-09-01',
      endDate: '2018-12-01',
      recruiterName: 'Alex Smith',
      recruiterEmail: 'alexsmith@example.org',
    })
    .jobPositionTitle({
      title: 'Job Title',
    })
    .jobPositionPurpose({
      purpose: 'Job purpose',
    })
    .jobPositionLocation({
      countryCode: 'SE',
      postalCode: '11356',
      municipality: '0180',
      addressLine: 'Birger Jarlsgatan 58, 11356, Stockholm',
      streetName: 'Birger Jarlsgatan 58',
    })
    .classification({
      scheduleType: 'part',
      duration: 'temporary',
      scheduleSummaryText: 'Schedule Summary',
      durationSummaryText: 'Duration Summary',
      termLength: 2,
    })
    .compensationDescription({
      currency: 'SEK',
      salaryType: 1,
      benefits: 'Benefits',
      summary: 'summary text',
    })
    .qualificationsRequiredSummary({
      summary: 'Summary of qualifications',
    })
    .qualification({
      type: 'license',
      description: 'DriversLicense',
      category: 'B',
    })
    .qualification({
      type: 'experience',
      required: true,
    })
    .qualification({
      type: 'equipment',
      description: 'Car',
    })
    .qualificationsPreferredSummary({
      summary: 'Preferred qualifications',
    })
    // applicationMethods() not neccessary, will be called by byWeb()
    // or byEmail(), included for clarity
    .applicationMethods()
    .byWeb({
      url: 'http://example.org',
    })
    .byEmail({
      email: 'foo@example.org',
    })
    .numberToFill({
      number: 1,
    })
    .hiringOrgDescription({
      description: 'Hiring org description',
    })
    .occupationGroup({
      code: 7652,
    });
} catch (err) {
  console.log(err);
  if (err.isJoi) {
    console.log('-----');
    console.log(err.annotate());
  }
  process.exit(1);
}
const xmlString = request.toString();
```

## Sample scripts
There are three scripts in the examples directory:
- minimal.js: demonstrate creation of XML for adding a new vacancy. Minimal fluff, start here.
- example.js: full fat demo, including adding, updating, and deleting vacancies.
See comments at top of script for more detail.
- apiTest.js: executes tests required by Arbetsförmedling to use API.*

\* Note that currently the API is returning an error for test #7 (add vacancy
  outside of Sweden), apparently on the address.

## Running the example script (./example/example.js)
The script uses [config](https://www.npmjs.com/package/config) to set organisation-specfic
information (company name, customer number, organisation number). If you just
want to see what the final XML looks like, you don't need to fill those in (but you
do need to create the configuration file):

```
cd example/config/

cp default.sample.json default.json

node example.js --create --xml
```

You can output the JSON object that is used to create the XML:

`node example.js --json`

You can use example.js to submit a test request. The API will return an error
unless valid values have been set in your config file:

`node example.js --create --submit`

You can set NODE_ENV to "production" if you want to submit to the live API.

## Oddities
Errors may be returned with HTTP status 202 or 400.

202 is returned with errors that appear to have been generated at the application level, e.g., from invalid values such as non-existent postal codes. Just because you got a 2xx response back doesn't mean the post was successful, check your response error code! (See example.js).

400 seems to usually be errors when validating against XSDs, although it will also be returned if you have an invalid customer number.

The JPPExtension element requires that children occur in a specific order. If you receive an error similar to:
```
  The element 'JPPExtension' in namespace
  'http://arbetsformedlingen.se/LedigtArbete' has
  invalid child element 'ApplicationReferenceID' in
  namespace
  'http://arbetsformedlingen.se/LedigtArbete'. List
  of possible elements expected:
  'LastDateApplication, ApplicationReferenceId,
  RequiredQualification, PreferredQualification' in
  namespace
  'http://arbetsformedlingen.se/LedigtArbete'.
```
check the order in which you are adding elements. Method calls must be in this order, though only occupationGroup() is required:
```
vacancy
  .hiringOrgDescription({
    description: 'Hiring org description',
  })
  .contact({
    name: 'Contact Name',
    phone: '555.555.5555',
    email: 'contact@example.org',
  })
  .occupationGroup({
    code: 7652,
  })
  .applicationReferenceID({
    id: 'ABC123',
  });
```

(Note that this appears to apply to other tag sets as well - the XSDs make liberal use of sequence - but this was the first time it bit me.)

There are two places where a description of the organization can be added. These are:
```
<Envelope><Packet><Payload><JobPositionPosting><HiringOrg><OrganizationalUnit><Description>
```
added via the `hiringOrg({name, id, url, description})` method, and
```
<Envelope><Packet><Payload><JPPExtension><HiringOrgDescription>
```
added via the `hiringOrgDescription({ description })` method.

Neither is required, and the documentation is unclear on the difference. Best guess until I try this out is that they appear in a different order in the job ad.

## Thank you
Development of this module was made possible by
[Internationella Engelska Skolan](https://engelska.se).

[@buren's ruby gem](https://github.com/buren/arbetsformedlingen) for creating
submissions was an invaluable resource. Also see that github page for
additional resources, including links to a Postman collection for querying
the Arbetsförmedlingen taxonomy service and data from that service in CSV
format.

## Contributing
Pull requests are welcome!

Methods implement a tag, or set of tags if they are logically a group. Each tag
or set is implemented with three methods:
- `jsonTagName()`: creates json for the tag/set,
- `validateTagName()`: performs parameter validation, throws Joi error on failure,
- `tagName()`: calls `validateTagName()` and `jsonTagName()`, performs any neccessary
business logic and attaches the new JSON object. Methods are chainable, so must
return `this`.

## Author
Dwayne Holmberg <dholmberg@othermachines.com>
[https://github.com/dwayneholmberg](https://github.com/dwayneholmberg)

## License
This module is freely distributable under the terms of the
[MIT license](https://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog
## 1.0.1 on 2018-03-19
- bugfix: timeStamp should not be hard coded.
## 1.0.0 on 2018-02-07
- BACKWARDS INCOMPATIBLE CHANGE. Calls to hiringOrgContact() must be removed, and contact information supplied in the call to `hiringOrg({ name, id, url, contact, description})`. See changelog note for 0.5.0.
## 0.5.0 on 2018-02-07
- IMPORTANT. This version is broken. HiringOrg > OrganizationalUnit > Description must occur *after* HiringOrg > Contact, or it will be rejected by the Arbetsförmedling's API. If you are using hiringOrgContact(), the tags will occur in the wrong order.
- added optional description parameter to hiringOrg() <HiriginOrg><OrganizationalUnit><Description>
## 0.4.0 on 2018-01-31
- add additional contacts for further information (JPPExtension > InformationContact)
- add a reference ID (JPPExtension > ApplicationReferenceId)
## 0.3.0 on 2018-01-16
- corrected validation for municipality codes.
- benefits and compensation description are optional, do not count on them being set when validating.
- benefits summary text is required.
- validation fixes for termLength and scheduleType in classification.
- update dependencies.

## 0.2.0 on 2017-11-27
- changed signature for byWeb(), should not accept summary text.
- added "required" attribute when adding an "experience required" qualifcation().
This is the same as "yearsOfExperience", but that attribute has a misleading name,
referring not to how many years of experience the employer desires, but whether
experience is required (4) or not required (1).
- many changes and additions to the example scripts, including examples for
updating and deleting vacancies and the beginning of script that performs
Arbetsförmedling's API tests.

## 0.1.4 on 2017-11-02
- Did not increment version number in package.json before pushing last commit. Get more coffee.

## 0.1.3 on 2017-11-02
- Updated README and CHANGELOG from 0.1.2 were not saved before pusblishing.
No functional changes in this update, this just corrects the documentation.

## 0.1.2 on 2017-11-02
- When adding a qualifcation(), create the parent element if it does not exist.
Previously, calling qualification() without first calling qualificationsRequired()
or qualificationsRequiredSummary() would result in an error.
- Correct package name in tests.
- Move release notes from README to CHANGELOG.

## 0.1.1 on 2017-10-07
- Updated meta information.

## 0.1.0 on 2017-10-07
- Initial release.


app.js
const Vacancy = require('./build/PlatsbankenVacancy.js');
const http = require('http');
const util = require('util');

const config = require('config');

const identity = config.get('identity');

// const hostname = 'localhost';
const hostname = 'api.arbetsformedlingen.se';
const port = 80;
const env = 'development';

// const baseUrl = 'http://api.arbetsformedlingen.se/ledigtarbete';
const baseUrl = '/ledigtarbete';
const routes = {
  postJobUrl: `${baseUrl}/apiledigtarbete/hrxml`,
  testPostJobUrl: `${baseUrl}/apiledigtarbete/test/hrxml`,
};

const path = env === 'production' ? routes.postJobUrl : routes.testPostJobUrl;

const options = {
  indent: '  ',
};

const request = Vacancy('http://arbetsformedlingen.se/LedigtArbete', '0.52', options);


try {
  request
    .sender({
      id: identity.customerNumber,
      email: 'foo@bar.com',
    })
    .transaction({
      id: 'ID HERE',
    })
    .jobPositionPosting({
      id: `${identity.orgNumber}-123`,
      status: 'active',
    })
    .hiringOrg({
      name: identity.companyName,
      id: identity.orgNumber,
      url: 'http://example.org',
    })
    .hiringOrgContact({
      countryCode: 'SE',
      postalCode: '11356',
      municipality: '0180',
      addressLine: 'Birger Jarlsgatan 58, 11356, Stockholm',
      streetName: 'Birger Jarlsgatan 58',
    })
    .postDetail({
      startDate: '2018-09-01',
      endDate: '2018-12-01',
      recruiterName: 'Alex Smith',
      recruiterEmail: 'alexsmith@example.org',
    })
    .jobPositionTitle({
      title: 'JOB TITLE',
    })
    .jobPositionPurpose({
      purpose: 'JOB PURPOSE',
    })
    .jobPositionLocation({
      countryCode: 'SE',
      postalCode: '11356',
      municipality: '0180',
      addressLine: 'Birger Jarlsgatan 58, 11356, Stockholm',
      streetName: 'Birger Jarlsgatan 58',
    })
    .classification({
      scheduleType: 'part',
      duration: 'temporary',
      scheduleSummaryText: 'Schedule Summary',
      durationSummaryText: 'Duration Summary',
      termLength: 2,
    })
    .compensationDescription({
      currency: 'SEK',
      salaryType: 1,
      benefits: 'bennies',
      summary: 'summary text',
    })
    .qualificationsRequiredSummary({
      summary: 'Summary of qualifications',
    })
    .qualification({
      type: 'license',
      description: 'DriversLicense',
      category: 'B',
    })
    .qualification({
      type: 'experience',
      yearsOfExperience: 1,
    })
    .qualification({
      type: 'equipment',
      description: 'Car',
    })
    .qualificationsPreferredSummary({
      summary: 'PREFERRED QUALIFICATIONS',
    })
    // applicationMethods() not neccessary, will be called by byWeb()
    // included for clarity
    .applicationMethods()
    .byWeb({
      url: 'http://example.org',
      summary: 'summary text',
    })
    .byEmail({
      email: 'foo@example.org',
    })
    .numberToFill({
      number: 1,
    })
    .hiringOrgDescription({
      description: 'HIRING ORG DESCRIPTION',
    })
    .occupationGroup({
      code: 7652,
    });
} catch (err) {
  console.log(err);
  console.log(err.message);
  console.log(err.details);
  if (err.isJoi) {
    console.log('-----');
    console.log(err.details[0].context);
    console.log('-----');
    console.log(err.annotate());
  }
  process.exit(1);
}
const xmlString = request.toString();
/*
console.log(util.inspect(request.doc, false, null, true));
console.log();
console.log('-'.repeat(80));
console.log();
console.log(`${request}`);
*/

console.log('-'.repeat(80));
console.log(`HOSTNAME: ${hostname}`);
console.log(`PATH: ${path}`);
console.log(`PORT: ${port}`);

console.log('-'.repeat(80));
console.log(xmlString);
console.log('-'.repeat(80));

process.exit();

const httpOptions = {
  hostname,
  port,
  path,
  method: 'POST',
  headers: {
    'Content-Type': 'text/xml',
    'Content-Length': xmlString.length,
  },
};

const req = http.request(httpOptions, (res) => {
  console.log(`STATUS: ${res.statusCode}`);
  console.log(`HEADERS: ${JSON.stringify(res.headers)}`);
  res.setEncoding('utf8');
  res.on('data', (chunk) => {
    console.log(`BODY: ${chunk}`);
  });

  res.on('end', () => {
    console.log('No more data in response.');
  });
});

req.on('error', (e) => {
  console.log(`problem with request: ${e.message}`);
});

// write data to request body
// xml would have been set somewhere to a complete xml document in the form of a string
req.write(xmlString);
req.end();


example/minimal.js
/*
* Minimal example for creating a new vacancy posting.
*
* Running this as-is with:
*   $ node minimal.js
*
* will output XML for adding a new vacancy. Using the -s flag
*   $ node minimal.js -s
*
* will submit that XML to Arbetsförmedling. This will return an error,
* complaining about a non-existent customer number, unless you fill in
* values for the identity object, just below.
*
* Full example of creating the XML for submission is in the createSubmission()
* function.
*/

const platsbankenVacancy = require('../build/platsbanken-vacancy.js');
const http = require('http');
const shortid = require('shortid');
const ArgumentParser = require('argparse').ArgumentParser;

const hostname = 'api.arbetsformedlingen.se';
const port = 80;
const path = '/ledigtarbete/apiledigtarbete/test/hrxml';

const identity = {
  customerNumber: '12345678',
  companyName: 'Company Name',
  orgNumber: '46-123456-1234',
};

const parser = new ArgumentParser({
  version: '0.0.1',
  addHelp: true,
  description: 'platsbanken-vacancy minimal example script',
});

parser.addArgument(
  ['-s', '--submit'],
  {
    help: 'submit postings to Arbetsförmedling',
    action: 'storeTrue',
  },
);

const args = parser.parseArgs();

const options = { indent: '  ' };

function createSubmission() {
  const vacancy = platsbankenVacancy('http://arbetsformedlingen.se/LedigtArbete', '0.52', options);

  vacancy
    .sender({
      id: identity.customerNumber,
      email: 'sender@example.com',
    })
    .transaction({
      id: shortid.generate(),
    })
    .jobPositionPosting({
      id: `${identity.orgNumber}-${shortid.generate()}`,
      status: 'active',
    })
    .hiringOrg({
      name: identity.companyName,
      id: identity.orgNumber,
      url: 'http://example.org/hiringOrg',
      contact: {
        countryCode: 'SE',
        postalCode: '11356',
        municipality: '0180',
        addressLine: 'Birger Jarlsgatan 58, 11356, Stockholm',
        streetName: 'Birger Jarlsgatan 58',
      },
      description: 'Organizational unit description',
    })
    .postDetail({
      startDate: '2018-09-01',
      endDate: '2018-12-01',
      recruiterName: 'Recruiter Name',
      recruiterEmail: 'recruiter@example.org',
    })
    .jobPositionTitle({
      title: 'Job Title',
    })
    .jobPositionPurpose({
      purpose: 'Job purpose',
    })
    .jobPositionLocation({
      countryCode: 'SE',
      postalCode: '11356',
      municipality: '0180',
      addressLine: 'Birger Jarlsgatan 58, 11356, Stockholm',
      streetName: 'Birger Jarlsgatan 58',
    })
    .classification({
      scheduleType: 'part',
      duration: 'temporary',
      scheduleSummaryText: 'Schedule Summary',
      durationSummaryText: 'Duration Summary',
      termLength: 2,
    })
    .compensationDescription({
      currency: 'SEK',
      salaryType: 1,
      benefits: 'Benefits',
      summary: 'Benefits summary text',
    })
    .qualificationsRequiredSummary({
      summary: 'Summary of qualifications',
    })
    .qualification({
      type: 'license',
      description: 'DriversLicense',
      category: 'B',
    })
    .qualification({
      type: 'experience',
      // same as: yearsOfExperience: 4
      required: true,
    })
    .qualification({
      type: 'equipment',
      description: 'Car',
    })
    .qualificationsPreferredSummary({
      summary: 'Preferred qualifications',
    })
    // applicationMethods() not neccessary, will be called by byWeb()
    // or byEmail(), included for clarity
    .applicationMethods()
    .byWeb({
      url: 'http://example.org/byWeb',
    })
    .byEmail({
      email: 'byEmailcontact@example.org',
    })
    .numberToFill({
      number: 1,
    })
    .hiringOrgDescription({
      description: 'Hiring org description',
    })
    .occupationGroup({
      code: 7652,
    });
  return vacancy;
}

try {
  vacancy = createSubmission();
} catch (err) {
  if (err.isJoi) {
    console.log('Validation error');
    console.log('-'.repeat(80));
    console.log(err);
    console.log('-'.repeat(80));
    console.log(err.annotate());
    console.log('-'.repeat(80));
  } else {
    console.log(err);
  }
  process.exit(1);
}

const payload = vacancy.toString();
console.log(payload);
console.log('-'.repeat(80));

if (args.submit) {
  const httpOptions = {
    hostname,
    port,
    path,
    method: 'POST',
    headers: {
      'Content-Type': 'text/xml',
    },
  };

  const req = http.request(httpOptions, (res) => {
    res.setEncoding('utf8');

    let body = '';
    res.on('data', (chunk) => {
      body += chunk;
    });

    res.on('end', () => {
      const response = JSON.parse(body);

      if (response[0] != null) {
        if (response[0].ErrorCode) {
          console.log('API Error');
          console.log(response[0].ErrorCode);
        }
        if (response[0].Message) {
          console.log(response[0].Message);
        }
      } else {
        if (response.ErrorCode) {
          console.log('API Error');
          console.log(response.ErrorCode);
        }
        if (response.Message) {
          console.log(response.Message);
        }
      }
    });
  });

  req.on('error', (e) => {
    console.log('Error');
    console.log(e.message);
    process.exit(1);
  });

  req.write(payload);
  req.end();
}


example/example.js
/*
* $ export NODE_ENV=production
*   Setting to 'production' will submit to the live server.
*   Defaults to 'development'.
*
* usage: example.js [-h] [-v] [-j] [-x] [-s] [-c] [-u] [-d] [-i ID]
*
* platsbanken-vacancy example script
*
* Optional arguments:
*  -h, --help      Show this help message and exit.
*  -v, --version   Show program's version number and exit.
*  -j, --json      output json
*  -x, --xml       output xml
*  -s, --submit    submit postings to Arbetsförmedling
*  -c, --create    create vacancy
*  -u, --update    update vacancy
*  -d, --delete    delete vacancy
*  -i ID, --id ID  posting id, will generate new one if not provided
*
* Examples
*   Unless you have set the NODE_ENV environment variable to "production",
*   these will submit to the test server:
*
*   output XML for adding a vacancy and exit (will not send):
*     $ node example.js --create --xml
*
*   submit a new vacancy to Arbetsförmedling:
*     $ node example.js --create --submit
*
*   update a vacancy at Arbetsförmedling:
*    $ node example.js --update --submit --id POST_ID_HERE
*/

const platsbankenVacancy = require('../build/platsbanken-vacancy.js');
const http = require('http');
const async = require('async');
const util = require('util');
const shortid = require('shortid');
const config = require('config');
const chalk = require('chalk');
const sprintf = require('sprintf-js').sprintf;
const vsprintf = require('sprintf-js').vsprintf;
const ArgumentParser = require('argparse').ArgumentParser;
const wordwrap = require('word-wrap');

const env = process.env.NODE_ENV === 'production' ? 'production' : 'development';
const identity = config.get('identity');

const hostname = 'api.arbetsformedlingen.se';
const port = 80;

const routes = {
  postJobLive: '/ledigtarbete/apiledigtarbete/hrxml',
  postJobTest: '/ledigtarbete/apiledigtarbete/test/hrxml',
};

const path = env === 'production' ? routes.postJobLive : routes.postJobTest;

const parser = new ArgumentParser({
  version: '0.0.1',
  addHelp: true,
  description: 'platsbanken-vacancy example script',
});

parser.addArgument(
  ['-j', '--json'],
  {
    help: 'output json',
    action: 'storeTrue',
  },
);

parser.addArgument(
  ['-x', '--xml'],
  {
    help: 'output xml',
    action: 'storeTrue',
  },
);

parser.addArgument(
  ['-s', '--submit'],
  {
    help: 'submit postings to Arbetsförmedling',
    action: 'storeTrue',
  },
);

parser.addArgument(
  ['-c', '--create'],
  {
    help: 'create vacancy',
    action: 'storeTrue',
  },
);

parser.addArgument(
  ['-u', '--update'],
  {
    help: 'update vacancy',
    action: 'storeTrue',
  },
);

parser.addArgument(
  ['-d', '--delete'],
  {
    help: 'delete vacancy',
    action: 'storeTrue',
  },
);

parser.addArgument(
  ['-i', '--id'],
  {
    help: 'posting id, will generate new one if not provided',
  },
);

const args = parser.parseArgs();

const options = { indent: '  ' };

/* start helper functions */

function createSubmission({ postingId, title, status }) {
  const vacancy = platsbankenVacancy('http://arbetsformedlingen.se/LedigtArbete', '0.52', options);

  vacancy
    .sender({
      id: identity.customerNumber,
      email: 'sender@example.com',
    })
    .transaction({
      id: shortid.generate(),
    })
    .jobPositionPosting({
      id: postingId,
      status,
    })
    .hiringOrg({
      name: identity.companyName,
      id: identity.orgNumber,
      url: 'http://example.org/hiringOrg',
      contact: {
        countryCode: 'SE',
        postalCode: '11356',
        municipality: '0180',
        addressLine: 'Birger Jarlsgatan 58, 11356, Stockholm',
        streetName: 'Birger Jarlsgatan 58',
      },
      description: 'Organizational unit description',
    })
    .postDetail({
      startDate: '2018-09-01',
      endDate: '2018-12-01',
      recruiterName: 'Recruiter Name',
      recruiterEmail: 'recruiter@example.org',
    })
    .jobPositionTitle({
      title,
    })
    .jobPositionPurpose({
      purpose: 'Job purpose',
    })
    .jobPositionLocation({
      countryCode: 'SE',
      postalCode: '11356',
      municipality: '0180',
      addressLine: 'Birger Jarlsgatan 58, 11356, Stockholm',
      streetName: 'Birger Jarlsgatan 58',
    })
    .classification({
      scheduleType: 'part',
      duration: 'temporary',
      scheduleSummaryText: 'Schedule Summary',
      durationSummaryText: 'Duration Summary',
      termLength: 2,
    })
    .compensationDescription({
      currency: 'SEK',
      salaryType: 1,
      benefits: 'Benefits',
      summary: 'Benefits summary text',
    })
    .qualificationsRequiredSummary({
      summary: 'Summary of qualifications',
    })
    .qualification({
      type: 'license',
      description: 'DriversLicense',
      category: 'B',
    })
    .qualification({
      type: 'experience',
      // same as: yearsOfExperience: 4
      required: true,
    })
    .qualification({
      type: 'equipment',
      description: 'Car',
    })
    .qualificationsPreferredSummary({
      summary: 'Preferred qualifications',
    })
    // applicationMethods() not neccessary, will be called by byWeb()
    // or byEmail(), included for clarity
    .applicationMethods()
    .byWeb({
      url: 'http://example.org/byWeb',
    })
    .byEmail({
      email: 'byEmailcontact@example.org',
    })
    .numberToFill({
      number: 1,
    })
    .hiringOrgDescription({
      description: 'Hiring org description',
    })
    .contact({
      name: 'Contact Name',
      phone: '555.555.5555',
      email: 'contact@example.org',
    })
    .contact({
      name: 'Contact Name2',
      phone: '555.555.5556',
      email: 'contact2@example.org',
    })
    .occupationGroup({
      code: 7652,
    })
    .applicationReferenceID({
      id: 'ABC123',
    });
  return vacancy;
}

function out(type, ...s) {
  // sprintf is just a placeholder for now. Will come back later and switch
  // to clui or similar when I have a moment. This isn't actually important
  // after all...
  if (s[1] == null) {
    s[1] = '';
  }

  // sprintf counts ascii codes, 10 for opening/closing a color
  const col2 = '%-49s  %29s';
  const col2color = '%-59s  %39s';
  if (type === 'info') {
    console.log(sprintf(col2color, chalk.cyan(s[0]), chalk.cyanBright(s[1])));
  } else if (type === 'success') {
    console.log(sprintf(col2color, chalk.green(s[0]), chalk.greenBright(s[1])));
  } else if (type === 'error') {
    console.log(sprintf(col2color, chalk.red(s[0]), chalk.redBright(s[1])));
  } else {
    console.log(vsprintf(col2, s));
  }
}

function send(data, next) {
  out('info', 'Sending...');
  const payload = data.vacancy.toString();

  const httpOptions = {
    hostname,
    port,
    path,
    method: 'POST',
    headers: {
      'Content-Type': 'text/xml',
    },
  };

  const req = http.request(httpOptions, (res) => {
    if (res.statusCode > 399) {
      out('error', 'Status', res.statusCode);
    } else {
      out('success', 'Status', res.statusCode);
    }

    res.setEncoding('utf8');

    let body = '';
    res.on('data', (chunk) => {
      body += chunk;
    });

    res.on('end', () => {
      const response = JSON.parse(body);

      if (response[0] != null) {
        if (response[0].ErrorCode) {
          out('error', 'API Error', response[0].ErrorCode);
        }
        if (response[0].Message) {
          console.log(wordwrap(response[0].Message));
        }
      } else {
        if (response.ErrorCode) {
          out('error', 'API Error', response.ErrorCode);
        }
        if (response.Message) {
          console.log(wordwrap(response.Message));
        }
      }
      next(null, data);
    });
  });

  req.on('error', (e) => {
    out('error', 'Error');
    console.log(e.message);
    next(e);
  });

  req.write(payload);
  req.end();
}

function processRequest(data, next) {
  if (args.json) {
    console.log(util.inspect(data.vacancy.doc, false, null, true));
  }

  if (args.xml) {
    console.log();
    console.log(data.vacancy.toString());
    console.log();
  }

  if (args.submit) {
    send(data, next);
  } else {
    next(null, data);
  }
}

function finalize(err, data) {
  if (err) {
    if (err.isJoi) {
      out('error', 'Validation error');
      console.log('-'.repeat(80));
      console.log(err);
      console.log('-'.repeat(80));
      console.log(err.annotate());
      console.log('-'.repeat(80));
    } else {
      console.log(err);
    }
    process.exit(1);
  } else {
    out('success', 'Done', data.postingId);
  }
}

/* end helper functions */

const tasks = [];
tasks.push((next) => {
  const postingId = args.id != null
    ? args.id
    : `${identity.orgNumber}-${shortid.generate()}`;

  next(null, { postingId });
});

if (args.create) {
  tasks.push((data, next) => {
    out('info', 'Creating new vacancy', data.postingId);

    data.vacancy = {};

    try {
      data.vacancy = createSubmission({
        postingId: data.postingId,
        title: 'Job title',
        status: 'active',
      });
    } catch (err) {
      next(err);
    }
    processRequest(data, next);
  });
}

if (args.update) {
  tasks.push((data, next) => {
    const postingId = args.id != null
      ? args.id
      : data.postingId;
    out('info', 'Updating vacancy', postingId);

    try {
      data.vacancy = createSubmission({
        postingId,
        title: 'Job title updated',
        status: 'active',
      });
    } catch (err) {
      next(err);
    }
    processRequest(data, next);
  });
}

if (args.delete) {
  tasks.push((data, next) => {
    const postingId = args.id != null
      ? args.id
      : data.postingId;
    out('info', 'Deleting vacancy', postingId);

    try {
      data.vacancy = createSubmission({
        postingId,
        title: 'Job title deleted',
        status: 'inactive',
      });
    } catch (err) {
      next(err);
    }
    processRequest(data, next);
  });
}

async.waterfall(tasks, finalize);
